CRM/Event/Form/Participant.php

```diff
@@ -2133,40 +2134,19 @@ protected function sendReceipts($params, array $participants, $lineItem, $additi
     }
 
     $customGroup = [];
-    //format submitted data
-    $customFieldsRole = [];
-    foreach ($this->getSubmittedValue('role_id') as $roleKey) {
-      $customFieldsRole = CRM_Utils_Array::crmArrayMerge(CRM_Core_BAO_CustomField::getFields('Participant',
-        FALSE, FALSE, $roleKey, $this->_roleCustomDataTypeID), $customFieldsRole);
-    }
-    $customFieldsEvent = CRM_Core_BAO_CustomField::getFields('Participant',
-      FALSE,
-      FALSE,
-      $this->getEventID(),
-      $this->_eventNameCustomDataTypeID
-    );
-    $customFieldsEventType = CRM_Core_BAO_CustomField::getFields('Participant',
-      FALSE,
-      FALSE,
-      $this->_eventTypeId,
-      $this->_eventTypeCustomDataTypeID
-    );
-    $customFields = CRM_Utils_Array::crmArrayMerge($customFieldsRole,
-      CRM_Core_BAO_CustomField::getFields('Participant', FALSE, FALSE, NULL, NULL, TRUE)
-    );
-    $customFields = CRM_Utils_Array::crmArrayMerge($customFieldsEvent, $customFields);
-    $customFields = CRM_Utils_Array::crmArrayMerge($customFieldsEventType, $customFields);
+    $customFieldFilters = [
+      'ParticipantRole' => $this->getSubmittedValue('role_id'),
+      'ParticipantEventName' => $this->getEventID(),
+      'ParticipantEventType' => $this->getEventValue('event_type_id'),
+    ];
+    $customFields = CRM_Core_BAO_CustomField::getViewableCustomFields('Participant', $customFieldFilters);
```

Comment:
This is a nasty bit of hard-coding that I've been working hard to kill off. It shouldn't be needed. APIv4 getfields now can give you this data (note the "grouping" tells you what value to map to):
![image](https://github.com/civicrm/civicrm-core/assets/2874912/85892393-bf3c-425a-b8f5-81dc431c6c80)

Reply:
@colemanw hmm - but I'm starting from an array of loaded Custom Fields & filtering it rather than doing an api call so I can't leverage that much AFAIK